### PR TITLE
Review Comments

### PR DESCRIPTION
**Pull request by xmas92**
Review comments for https://github.com/openjdk/lilliput/pull/181

### CODE REVIEW

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -297,53 +297,57 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
       cmp(zr, obj);
       b(slow_path);
     } else {
+      if (UseObjectMonitorTable) {
```
Comment (xmas92) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp]: `aarch64` always did a cache lookup even when `!UseObjectMonitorTable` which always results in a slow path.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -6780,8 +6780,10 @@ void MacroAssembler::lightweight_lock(Register basic_lock, Register obj, Registe
   // instruction emitted as it is part of C1's null check semantics.
   ldr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
 
-  // Clear cache in case fast locking succeeds.
-  str(zr, Address(basic_lock, BasicObjectLock::lock_offset() + in_ByteSize((BasicLock::object_monitor_cache_offset_in_bytes()))));
+  if (UseObjectMonitorTable) {
```
Comment (xmas92) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:6783]: Clearing not needed when `!UseObjectMonitorTable`

File: src/hotspot/share/oops/markWord.cpp [hunk 3]
```diff
@@ -76,7 +76,7 @@ void markWord::print_on(outputStream* st, bool print_monitor_info) const {
   } else if (has_monitor()) {  // last bits = 10
     // have to check has_monitor() before is_locked()
     st->print(" monitor(" INTPTR_FORMAT ")=", value());
-    if (print_monitor_info && LockingMode != LM_LIGHTWEIGHT) {
+    if (print_monitor_info && !UseObjectMonitorTable) {
```
Comment (xmas92) [inline src/hotspot/share/oops/markWord.cpp:79]: Only with `UseObjectMonitorTable` do we not have a monitor in the markWord.

File: src/hotspot/share/oops/markWord.hpp [hunk 4]
```diff
@@ -240,8 +240,11 @@ class markWord {
   }
   bool has_displaced_mark_helper() const {
     intptr_t lockbits = value() & lock_mask_in_place;
-    return LockingMode == LM_LIGHTWEIGHT  ? UseObjectMonitorTable ? false : lockbits == monitor_value   // no displace mark or monitor?
-                                         : (lockbits & unlocked_value) == 0; // monitor | stack-locked?
+    if (LockingMode == LM_LIGHTWEIGHT) {
```
Comment (xmas92) [inline src/hotspot/share/oops/markWord.hpp:243]: Take this if you want. There were some extra whitespaces, and the nested ternaries are confusing at best.

File: src/hotspot/share/opto/library_call.cpp [hunk 5]
```diff
@@ -4578,21 +4578,23 @@ bool LibraryCallKit::inline_native_hashcode(bool is_virtual, bool is_static) {
...
-    Node *monitor_val   = _gvn.MakeConX(markWord::monitor_value);
-    Node *chk_monitor   = _gvn.transform(new CmpXNode(lmasked_header, monitor_val));
-    Node *test_monitor  = _gvn.transform(new BoolNode(chk_monitor, BoolTest::eq));
-
-    generate_slow_guard(test_monitor, slow_region);
-  } else {
-    Node *unlocked_val      = _gvn.MakeConX(markWord::unlocked_value);
-    Node *chk_unlocked      = _gvn.transform(new CmpXNode(lmasked_header, unlocked_val));
-    Node *test_not_unlocked = _gvn.transform(new BoolNode(chk_unlocked, BoolTest::ne));
+  if (!UseObjectMonitorTable) {
```
Comment (xmas92) [inline src/hotspot/share/opto/library_call.cpp:4581]: With `UseObjectMonitorTable` we do not have to guards for any locking bits.

File: src/hotspot/share/runtime/arguments.cpp [hunk 6]
```diff
@@ -2949,12 +2949,12 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
   UNSUPPORTED_OPTION(ShowRegistersOnAssert);
 #endif // CAN_SHOW_REGISTERS_ON_ASSERT
 
-#ifdef _LP64
   if (UseObjectMonitorTable && LockingMode != LM_LIGHTWEIGHT) {
```
Comment (xmas92) [inline src/hotspot/share/runtime/arguments.cpp:2952]: This is still required for 32bit

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ObjectSynchronizer.java [hunk 7]
```diff
@@ -55,7 +55,7 @@ public long identityHashValueFor(Oop obj) {
       // FIXME: can not generate marks in debugging system
       return mark.hash();
     } else if (mark.hasMonitor()) {
-      if (VM.getVM().getCommandLineFlag("LockingMode").getInt() == LockingMode.getLightweight()) {
+      if (VM.getVM().getCommandLineFlag("UseObjectMonitorTable").getBool()) {
```
Comment (xmas92) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ObjectSynchronizer.java:58]: Only with `UseObjectMonitorTable` can we get the hash from the markWord when it has a monitor.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Mark.java [hunk 8]
```diff
@@ -172,7 +172,7 @@ public ObjectMonitor monitor() {
     if (Assert.ASSERTS_ENABLED) {
       Assert.that(hasMonitor(), "check");
     }
-    if (VM.getVM().getCommandLineFlag("LockingMode").getInt() == LockingMode.getLightweight()) {
+    if (VM.getVM().getCommandLineFlag("UseObjectMonitorTable").getBool()) {
```
Comment (xmas92) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Mark.java:175]: The previous method worked, but is only required with `UseObjectMonitorTable`

**Comment by xmas92** [comment]
Changed `OM_OFFSET_NO_MONITOR_VALUE_TAG` to depend on `UseObjectMonitorTable` instead because it is still used by s390, otherwise it breaks that platform. I also saw ~2% regression in the inflation µbenchmarks when untagging the monitor. 

In general reverted the C2 behaviour when `UseObjectMonitorTable` is false to emit the same code as before this patch. 

The `OM_OFFSET_NO_MONITOR_VALUE_TAG` is pretty ugly and misleading so will attempt to cleanup the usage as well. 
Also the `OMUseC2Cache` is unwieldily and is something I was going to remove next anyways. I will added it to this review PR as well.

**Comment by coleenp** [comment]
> Changed OM_OFFSET_NO_MONITOR_VALUE_TAG to depend on UseObjectMonitorTable instead because it is still used by s390, otherwise it breaks that platform. I also saw ~2% regression in the inflation µbenchmarks when untagging the monitor.

This 2% regression is really unfortunate because this OM_OFFSET_NO_MONITOR_VALUE_TAG is so ugly.

Thanks for all these changes and corrections.  I was sure I missed several things.  I'll ask you offline how to merge this in.

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp]: I see that now.  Thanks for removing OMUseC2Cache.

**Review by coleenp — APPROVED** [review]

**Pull request merged**